In [ ]:
!python -m pip install -U python-woc pandas matplotlib
# please clear the output of this cell in your notebook before checking it in

In [4]:
import pandas as pd, time
from tqdm import tqdm
from woc.remote import WocMapsRemote

sel = pd.DataFrame([
    ('lsstdesc_weaklensingdeblending',  'https://github.com/LSSTDESC/WeakLensingDeblending'),
    ('mkazhdan_poissonrecon',           'https://github.com/mkazhdan/PoissonRecon'),
    ('julianlsolvers_nlsolversbase.jl', 'https://github.com/JuliaNLSolvers/NLSolversBase.jl'),
    ('gridtools_gridtools',             'https://github.com/GridTools/gridtools'),
    ('christophsax_seasonal',           'https://github.com/christophsax/seasonal'),
    ('svf-tools_svf',                   'https://github.com/SVF-tools/SVF'),
    ('stephenslab_flashr',              'https://github.com/stephenslab/flashr'),
    ('feihoo87_qulab',                  'https://github.com/feihoo87/QuLab'),
    ('soerenpannier_emdi',              'https://github.com/SoerenPannier/emdi'),
    ('gimli-org_gimli',                 'https://github.com/gimli-org/gimli'),
], columns=['wocid', 'url'])

woc = WocMapsRemote(base_url="https://worldofcode.org/api/")

list_df = []
for row in sel.itertuples():
    try:
        shas = woc.get_values('p2c', row.wocid)
    except Exception as e:
        print('FAILED', row.wocid, e); continue
    print(row.wocid, len(shas))
    d = pd.DataFrame(shas, columns=['sha1']); d['project'] = row.wocid
    list_df.append(d)
    time.sleep(1)

df = pd.concat(list_df, ignore_index=True)
df.to_csv('df_commits.csv', index=False)

lsstdesc_weaklensingdeblending 735
mkazhdan_poissonrecon 414
julianlsolvers_nlsolversbase.jl 480
gridtools_gridtools 16440
christophsax_seasonal 1227
svf-tools_svf 6530
stephenslab_flashr 616
feihoo87_qulab 4361
soerenpannier_emdi 1143
gimli-org_gimli 7466


In [13]:
import os
os.makedirs('chunks', exist_ok=True)

for prj, g in df.groupby('project'):
    fn = f'chunks/{prj}.csv'
    if os.path.exists(fn):
        print('skipping (done):', prj); continue
    shas = g['sha1'].tolist()
    rows = []
    for i in tqdm(range(0, len(shas), 10), desc=prj):
        chunk = shas[i:i+10]
        for attempt in range(3):                      # retry up to 3 times
            try:
                res, err = woc.get_values_many('commit.tch', chunk)
                if err: print('Got Errors', err)
                res = {k: v[0] for k, v in res.items()}
                break
            except Exception as e:
                print('retry', attempt, e); time.sleep(5)
        else:
            print('GAVE UP on chunk in', prj); continue
        for sha, c in res.items():
            rows.append({'project_wocid': prj, 'commit_sha1': sha,
                         'author': c[2][0], 'time': int(c[2][1]),
                         'commit message': c[4]})
        time.sleep(1)
    print(prj, 'expected', len(shas), 'got', len(rows))
    pd.DataFrame(rows).to_csv(fn, index=False)

christophsax_seasonal: 100%|█████████████████████████████████████████████████████████| 123/123 [02:06<00:00,  1.03s/it]


christophsax_seasonal expected 1227 got 1227


feihoo87_qulab:  30%|███████████████████▏                                            | 131/437 [02:15<05:17,  1.04s/it]

Got Errors {'4df6f45fb773ecc76f1d59b1ab0cf6306911f296': 'Key 4df6f45fb773ecc76f1d59b1ab0cf6306911f296 not found in /da5_fast/All.sha1c/commit_77.tch'}


feihoo87_qulab: 100%|████████████████████████████████████████████████████████████████| 437/437 [07:34<00:00,  1.04s/it]


feihoo87_qulab expected 4361 got 4360


gimli-org_gimli:   3%|█▋                                                              | 19/747 [00:19<12:39,  1.04s/it]

Got Errors {'0621982adee96d22bcd2b67adea06e29818e8a21': 'Key 0621982adee96d22bcd2b67adea06e29818e8a21 not found in /da5_fast/All.sha1c/commit_6.tch'}


gimli-org_gimli:  13%|████████▎                                                       | 97/747 [01:40<11:14,  1.04s/it]

Got Errors {'21a63bfd6c3d66605e75303f25892a1e632594d1': 'Key 21a63bfd6c3d66605e75303f25892a1e632594d1 not found in /da5_fast/All.sha1c/commit_33.tch'}


gimli-org_gimli:  20%|████████████▍                                                  | 148/747 [02:33<10:24,  1.04s/it]

Got Errors {'3328c3a33c11577be82a46d6ca5ae6701492914f': 'Key 3328c3a33c11577be82a46d6ca5ae6701492914f not found in /da5_fast/All.sha1c/commit_51.tch'}


gimli-org_gimli:  25%|███████████████▉                                               | 189/747 [03:16<09:39,  1.04s/it]

Got Errors {'426279b084c45857b71832e580413d192cf8cb5e': 'Key 426279b084c45857b71832e580413d192cf8cb5e not found in /da5_fast/All.sha1c/commit_66.tch'}


gimli-org_gimli:  46%|█████████████████████████████▏                                 | 346/747 [05:58<06:54,  1.03s/it]

Got Errors {'77c44a3c87b6fee71752cc1f57e7683f4a836eaf': 'Key 77c44a3c87b6fee71752cc1f57e7683f4a836eaf not found in /da5_fast/All.sha1c/commit_119.tch'}


gimli-org_gimli:  46%|█████████████████████████████▎                                 | 347/747 [06:00<06:53,  1.03s/it]

Got Errors {'7825ff5a4e010a6b756a5725612c9a20f16b0548': 'Key 7825ff5a4e010a6b756a5725612c9a20f16b0548 not found in /da5_fast/All.sha1c/commit_120.tch', '782abd05bd5c0e02f60f5e1ccc9182c2f688763f': 'Key 782abd05bd5c0e02f60f5e1ccc9182c2f688763f not found in /da5_fast/All.sha1c/commit_120.tch'}


gimli-org_gimli:  50%|███████████████████████████████▎                               | 371/747 [06:24<06:28,  1.03s/it]

Got Errors {'80297a1a87aca6288a3d8fa21727c17f6fa4d5eb': 'Key 80297a1a87aca6288a3d8fa21727c17f6fa4d5eb not found in /da5_fast/All.sha1c/commit_0.tch'}


gimli-org_gimli:  84%|█████████████████████████████████████████████████████▏         | 631/747 [10:52<01:59,  1.03s/it]

Got Errors {'d8d66333237ffae89d610beb76592b6b92844b46': 'Key d8d66333237ffae89d610beb76592b6b92844b46 not found in /da5_fast/All.sha1c/commit_88.tch'}


gimli-org_gimli:  92%|██████████████████████████████████████████████████████████▏    | 690/747 [11:53<00:58,  1.03s/it]

Got Errors {'ed0c3a6902ee05e57b107126621906ee686a932c': 'Key ed0c3a6902ee05e57b107126621906ee686a932c not found in /da5_fast/All.sha1c/commit_109.tch'}


gimli-org_gimli:  93%|██████████████████████████████████████████████████████████▎    | 692/747 [11:55<00:56,  1.03s/it]

Got Errors {'edb612e1d1a8c92cb20c12121e188a256db5e419': 'Key edb612e1d1a8c92cb20c12121e188a256db5e419 not found in /da5_fast/All.sha1c/commit_109.tch'}


gimli-org_gimli:  99%|██████████████████████████████████████████████████████████████▍| 741/747 [12:46<00:06,  1.03s/it]

Got Errors {'fe34b34252344b47842897920b98d73e6c6e162e': 'Key fe34b34252344b47842897920b98d73e6c6e162e not found in /da5_fast/All.sha1c/commit_126.tch'}


gimli-org_gimli: 100%|███████████████████████████████████████████████████████████████| 747/747 [12:52<00:00,  1.03s/it]


gimli-org_gimli expected 7466 got 7454


gridtools_gridtools:   5%|██▋                                                        | 74/1644 [01:16<27:01,  1.03s/it]

Got Errors {'0ba333499b983b1607bca190db6f3ae84b06c15d': 'Key 0ba333499b983b1607bca190db6f3ae84b06c15d not found in /da5_fast/All.sha1c/commit_11.tch'}


gridtools_gridtools:  41%|███████████████████████▍                                  | 666/1644 [11:28<16:54,  1.04s/it]

Got Errors {'699a5f8b8a2e2d6d4c8246eb70c6b8b3e4216043': 'Key 699a5f8b8a2e2d6d4c8246eb70c6b8b3e4216043 not found in /da5_fast/All.sha1c/commit_105.tch'}


gridtools_gridtools: 100%|█████████████████████████████████████████████████████████| 1644/1644 [28:18<00:00,  1.03s/it]


gridtools_gridtools expected 16440 got 16438


julianlsolvers_nlsolversbase.jl:  10%|█████▏                                            | 5/48 [00:05<00:44,  1.03s/it]

Got Errors {'1cfca0050c3faaf08c25eaadfdae685ded0729d5': 'Key 1cfca0050c3faaf08c25eaadfdae685ded0729d5 not found in /da5_fast/All.sha1c/commit_28.tch'}


julianlsolvers_nlsolversbase.jl:  48%|███████████████████████▍                         | 23/48 [00:23<00:25,  1.04s/it]

Got Errors {'78b6b4175d804abbc373f9da8e1c0cb343f4a9d6': 'Key 78b6b4175d804abbc373f9da8e1c0cb343f4a9d6 not found in /da5_fast/All.sha1c/commit_120.tch'}


julianlsolvers_nlsolversbase.jl:  88%|██████████████████████████████████████████▉      | 42/48 [00:43<00:06,  1.03s/it]

Got Errors {'e41d89a7eef874202d2c38d735c3225d0980c019': 'Key e41d89a7eef874202d2c38d735c3225d0980c019 not found in /da5_fast/All.sha1c/commit_100.tch'}


julianlsolvers_nlsolversbase.jl:  98%|███████████████████████████████████████████████▉ | 47/48 [00:48<00:01,  1.03s/it]

Got Errors {'fa5c8dab9103e4b5feb658016e1791b4b567f450': 'Key fa5c8dab9103e4b5feb658016e1791b4b567f450 not found in /da5_fast/All.sha1c/commit_122.tch'}


julianlsolvers_nlsolversbase.jl: 100%|█████████████████████████████████████████████████| 48/48 [00:49<00:00,  1.03s/it]


julianlsolvers_nlsolversbase.jl expected 480 got 476


lsstdesc_weaklensingdeblending: 100%|██████████████████████████████████████████████████| 74/74 [01:16<00:00,  1.03s/it]


lsstdesc_weaklensingdeblending expected 735 got 735


mkazhdan_poissonrecon: 100%|███████████████████████████████████████████████████████████| 42/42 [00:43<00:00,  1.04s/it]


mkazhdan_poissonrecon expected 414 got 414


soerenpannier_emdi: 100%|████████████████████████████████████████████████████████████| 115/115 [01:59<00:00,  1.04s/it]


soerenpannier_emdi expected 1143 got 1143


stephenslab_flashr:   0%|                                                                       | 0/62 [00:00<?, ?it/s]

Got Errors {'04003c3657d72f87016108b318accee18a30d274': 'Key 04003c3657d72f87016108b318accee18a30d274 not found in /da5_fast/All.sha1c/commit_4.tch'}


stephenslab_flashr:  47%|█████████████████████████████                                 | 29/62 [00:29<00:34,  1.03s/it]

Got Errors {'768e8664491f5df26e4c29d310965351ce1217d5': 'Key 768e8664491f5df26e4c29d310965351ce1217d5 not found in /da5_fast/All.sha1c/commit_118.tch'}


stephenslab_flashr: 100%|██████████████████████████████████████████████████████████████| 62/62 [01:03<00:00,  1.03s/it]


stephenslab_flashr expected 616 got 614


svf-tools_svf:  21%|█████████████▋                                                   | 138/653 [02:23<08:54,  1.04s/it]

Got Errors {'37b5dc1ca5d4a7fbb040e42887e87d6e976e4dc5': 'Key 37b5dc1ca5d4a7fbb040e42887e87d6e976e4dc5 not found in /da5_fast/All.sha1c/commit_55.tch'}


svf-tools_svf:  35%|██████████████████████▉                                          | 231/653 [04:00<07:22,  1.05s/it]

Got Errors {'5bf807179867e08e30eb9760842fd094d822c2f1': 'Key 5bf807179867e08e30eb9760842fd094d822c2f1 not found in /da5_fast/All.sha1c/commit_91.tch'}


svf-tools_svf:  83%|█████████████████████████████████████████████████████▊           | 541/653 [09:24<01:56,  1.04s/it]

Got Errors {'d5d8128d8d54cb037ad17be6afd4b8783c32e258': 'Key d5d8128d8d54cb037ad17be6afd4b8783c32e258 not found in /da5_fast/All.sha1c/commit_85.tch'}


svf-tools_svf: 100%|█████████████████████████████████████████████████████████████████| 653/653 [11:21<00:00,  1.04s/it]

svf-tools_svf expected 6530 got 6527


In [6]:
import os
out = pd.concat([pd.read_csv(f'chunks/{f}') for f in sorted(os.listdir('chunks'))],
                ignore_index=True)
out.to_csv('tcartier_project_summary.csv', index=False, sep=';')
print(len(out), 'rows')
out.head(3)

39388 rows


,project_wocid,commit_sha1,author,time,commit message
0,christophsax_seasonal,00352d722b0b33cc67cf506491555f1c455297bd,Christoph Sax <christoph.sax@gmail.com>,1600417136,Merge pull request #253 from christophsax/stat...
1,christophsax_seasonal,003d9dd00c851c838f2c688bb21116da19dfad62,Christoph <christoph.sax@gmail.com>,1385979532,update hidden extension remark\n
2,christophsax_seasonal,0053e34f6fcce45cb28508877d0c6f918b0fd7e9,Christoph Sax <christoph.sax@gmail.com>,1600414334,"rm_defaults, resolves #251\n"


In [7]:
chk = pd.read_csv('tcartier_project_summary.csv', sep=';')
print(chk.columns.tolist())   # ['project_wocid','commit_sha1','author','time','commit message']
print(len(chk), 'rows')       # should match len(out)
print(chk['project_wocid'].nunique(), 'projects')   # should be 10
print(chk.isna().sum())       # a few empty messages are fine

['project_wocid', 'commit_sha1', 'author', 'time', 'commit message']
39388 rows
10 projects
project_wocid      0
commit_sha1        0
author             0
time               0
commit message    31
dtype: int64


In [8]:
!ls -lh tcartier_project_summary.csv

-rw-------. 1 tcartier tcartier 6.2M Sep 29 14:41 tcartier_project_summary.csv


In [9]:
out['email'] = out['author'].str.extract(r'<(.+?)>')
out['dt'] = pd.to_datetime(out['time'], unit='s')
woc_summary = out.groupby('project_wocid').agg(
    commits=('commit_sha1', 'nunique'),
    authors_by_string=('author', 'nunique'),
    authors_by_email=('email', 'nunique'),
    min_time=('dt', 'min'),
    max_time=('dt', 'max'))
woc_summary

,commits,authors_by_string,authors_by_email,min_time,max_time
project_wocid,,,,,
christophsax_seasonal,1227,17,16,2013-09-17 20:14:42,2025-02-28 17:18:20
feihoo87_qulab,4360,11,10,2014-12-09 11:09:39,2025-09-08 12:17:40
gimli-org_gimli,7454,77,63,2011-05-06 17:56:03,2025-11-03 19:23:44
gridtools_gridtools,16438,106,90,2013-10-25 08:01:33,2026-03-03 10:45:08
julianlsolvers_nlsolversbase.jl,476,47,45,2016-01-26 10:00:10,2025-06-26 02:32:58
lsstdesc_weaklensingdeblending,735,26,22,2013-03-11 11:27:58,2026-01-16 19:46:14
mkazhdan_poissonrecon,414,58,48,2015-07-14 23:31:43,2025-10-25 01:00:38
soerenpannier_emdi,1143,23,19,2016-10-28 12:35:37,2025-09-05 12:35:30
stephenslab_flashr,614,7,5,2017-09-12 19:03:25,2024-12-04 13:11:19


In [11]:
import requests, pandas as pd

repos = {
 'lsstdesc_weaklensingdeblending':  'LSSTDESC/WeakLensingDeblending',
 'mkazhdan_poissonrecon':           'mkazhdan/PoissonRecon',
 'julianlsolvers_nlsolversbase.jl': 'JuliaNLSolvers/NLSolversBase.jl',
 'gridtools_gridtools':             'GridTools/gridtools',
 'christophsax_seasonal':           'christophsax/seasonal',
 'svf-tools_svf':                   'SVF-tools/SVF',
 'stephenslab_flashr':              'stephenslab/flashr',
 'feihoo87_qulab':                  'feihoo87/QuLab',
 'soerenpannier_emdi':              'SoerenPannier/emdi',
 'gimli-org_gimli':                 'gimli-org/gimli',
}
headers = {}   # or {"Authorization": "Bearer YOUR_TOKEN"} (don't commit the token)

gh = []
for wocid, o_r in repos.items():
    r = requests.get(f'https://api.github.com/repos/{o_r}', headers=headers).json()
    c = requests.get(f'https://api.github.com/repos/{o_r}/commits?per_page=1', headers=headers).json()
    gh.append({'project': wocid, 'stars': r['stargazers_count'],
               'forks': r['forks_count'], 'last_commit': c[0]['commit']['committer']['date']})
gh_df = pd.DataFrame(gh)
gh_df

,project,stars,forks,last_commit
0,lsstdesc_weaklensingdeblending,15,15,2025-07-24T20:47:53Z
1,mkazhdan_poissonrecon,1860,470,2026-04-29T15:37:55Z
2,julianlsolvers_nlsolversbase.jl,34,31,2026-08-17T16:50:53Z
3,gridtools_gridtools,68,24,2026-02-18T13:11:51Z
4,christophsax_seasonal,120,30,2026-09-17T18:04:00Z
5,svf-tools_svf,1711,498,2026-09-29T11:52:00Z
6,stephenslab_flashr,54,11,2024-12-04T13:11:19Z
7,feihoo87_qulab,18,20,2026-07-14T15:00:58Z
8,soerenpannier_emdi,17,24,2025-08-23T17:57:56Z
9,gimli-org_gimli,509,163,2026-09-24T10:31:28Z


### GitHub stats (retrieved Sept 29, 2026)
| project | stars | forks | last commit |
|---|---|---|---|
| lsstdesc_weaklensingdeblending | 15 | 15 | 2025-07-24 |
| mkazhdan_poissonrecon | 1860 | 470 | 2026-04-29 |
| julianlsolvers_nlsolversbase.jl | 34 | 31 | 2026-08-17 |
| gridtools_gridtools | 68 | 24 | 2026-02-18 |
| christophsax_seasonal | 120 | 30 | 2026-09-17 |
| svf-tools_svf | 1711 | 498 | 2026-09-29 |
| stephenslab_flashr | 54 | 11 | 2024-12-04 |
| feihoo87_qulab | 18 | 20 | 2026-07-14 |
| soerenpannier_emdi | 17 | 24 | 2025-08-23 |
| gimli-org_gimli | 509 | 163 | 2026-09-24 |

### WoC stats
| project | commits | authors (by email) | authors (by name string) | min time | max time |
|---|---|---|---|---|---|
| christophsax_seasonal | 1227 | 16 | 17 | 2013-09-17 | 2025-02-28 |
| feihoo87_qulab | 4360 | 10 | 11 | 2014-12-09 | 2025-09-08 |
| gimli-org_gimli | 7454 | 63 | 77 | 2011-05-06 | 2025-11-03 |
| gridtools_gridtools | 16438 | 90 | 106 | 2013-10-25 | 2026-03-03 |
| julianlsolvers_nlsolversbase.jl | 476 | 45 | 47 | 2016-01-26 | 2025-06-26 |
| lsstdesc_weaklensingdeblending | 735 | 22 | 26 | 2013-03-11 | 2026-01-16 |
| mkazhdan_poissonrecon | 414 | 48 | 58 | 2015-07-14 | 2025-10-25 |
| soerenpannier_emdi | 1143 | 19 | 23 | 2016-10-28 | 2025-09-05 |
| stephenslab_flashr | 614 | 5 | 7 | 2017-09-12 | 2024-12-04 |
| svf-tools_svf | 6527 | 196 | 238 | 2015-06-05 | 2025-11-06 |

Notes: 24 of 39,412 commits (0.06%) couldn't be retrieved after retries. Author counts by name string overcount because the same person shows up with name variants. WoC data extends beyond the Jan 2025 cutoff in the handout. For gridtools, weaklensingdeblending, and emdi, WoC's latest commit is later than GitHub's last commit, possibly from non-default branches/forks or author vs committer time.

# Make sure you check in to your fork not just the notebook but also the csv files!!!

# Don't forget to add requested data from github and this notebook
### For each of the 10 projects go to their github repo and get the number of stars, number of forks, and the last commit date
### Report (in your notebook) the number of commits, the number of authors, and max and min time for each project based on WoC commits and also add the info you obtained from github